pageserver/src/tenant/config.rs

```diff
@@ -44,7 +45,225 @@ pub mod defaults {
+        // TODO: add an override for disk pressure in AttachedLocationConfig,
+        // and respect it here.
+        match &self.mode {
+            LocationMode::Attached(attach_conf) => {
+                match attach_conf.attach_mode {
+                    AttachmentMode::Single => true,
+                    AttachmentMode::Multi | AttachmentMode::Stale => {
+                        // In Multi mode we avoid doing deletions because some other
+                        // attached pageserver might get 404 while trying to read
+                        // a layer we delete which is still referenced in their metadata.
+                        //
+                        // In Stale mode, we avoid doing deletions because we expect
+                        // that they would ultimately fail validation in the deletion
+                        // queue due to our stale generation.
+                        false
+                    }
+                }
+            }
+            LocationMode::Secondary(_) => {
+                // Do not expect to be called in this state
+                tracing::error!("Called may_delete_layers_hint on a tenant in secondary mode");
+                false
+            }
+        }
+    }
+
+    /// Whether we are currently hinted that it is worthwhile to upload layers.
+    /// This is only advisory, not required for data safety.
+    /// See [`AttachmentMode`] for more context.
+    pub(crate) fn may_upload_layers_hint(&self) -> bool {
+        // TODO: add an override for disk pressure in AttachedLocationConfig,
+        // and respect it here.
+        match &self.mode {
+            LocationMode::Attached(attach_conf) => {
+                match attach_conf.attach_mode {
+                    AttachmentMode::Single | AttachmentMode::Multi => true,
+                    AttachmentMode::Stale => {
+                        // In Stale mode, we avoid dong uploads because we expect that
+                        // our replacement pageserver will already have started its own
+                        // IndexPart that will never reference layers we upload: it is
+                        // wasteful.
+                        false
+                    }
+                }
+            }
+            LocationMode::Secondary(_) => {
+                // Do not expect to be called in this state
+                tracing::error!("Called may_upload_layers_hint on a tenant in secondary mode");
+                false
+            }
+        }
+    }
+}
+
+impl TryFrom<&'_ models::LocationConfig> for LocationConf {
+    type Error = anyhow::Error;
+
+    fn try_from(conf: &'_ models::LocationConfig) -> Result<Self, Self::Error> {
+        let tenant_conf = TenantConfOpt::try_from(&conf.tenant_conf)?;
+
+        fn get_generation(conf: &'_ models::LocationConfig) -> Result<Generation, anyhow::Error> {
+            conf.generation
+                .ok_or_else(|| anyhow::anyhow!("Generation must be set when attaching"))
+        }
```

Comment:
Nothing about TryFrom says it's only used in attaching context, but the error and checks are specific to that use case.
Also, the only user of this impl doesn't need it to be `try_from``.
It think this should just be a regular static method on `LocationConf`.

Also, IMO, we should assert absence of the `generation` field in the cases where we don't `get_generation()`.
And, I'm wondering whether for safety, the API models for LocationConfig should be serde `deny_unknown_fields`.
Would rather bail out than silently handle a mismatch betweenn control plane and pageserver mental model.
We can prepare for backward-compatibility by putting a `version` field in the struct (or, in the future, use an API prefix != `/v1`, but that's a big pain.)

Reply:
> Nothing about TryFrom says it's only used in attaching context, but the error and checks are specific to that use case.

The "Generation must be set when attaching" error is correct here: the `get_generation` function is used in the paths in the match below where we are using one of the attached states.

> Also, the only user of this impl doesn't need it to be try_from``. It think this should just be a regular static method on LocationConf`.

Hmm, interesting style question.  My habit is to use such a trait if what I'm doing aligns with it -- when I'm writing a function that converts one type to another but can throw an error, that's a TryFrom, although you're quite right that it doesn't have to be.  Don't feel strongly.  I'll move the try_from method into the main impl block, since that's a bit fewer LoC.

> Also, IMO, we should assert absence of the generation field in the cases where we don't get_generation().

Good idea, added for the secondary state.

> We can prepare for backward-compatibility by putting a version field in the struct (or, in the future, use an API prefix != /v1, but that's a big pain.)

Hmm.  I think I would struggle to justify why all the other API endpoints just get a version in the path, but the location configuration API gets one in the struct.  Either way, it requires the caller to be version aware and include one.  I think if I needed to make a backward incompatible change, I would either let the caller provide a config version as a URL parameter, or just replace the /v1/ with /v2/ (it's not that onerous if the old version is only around for one deploy cycle).

Changes for the points above in this function are a6453b09a